# Einführung in Python für die Computational Social Science (CSS)
## Session 04 - Pandas



## Jonas Volle
Wissenschaftlicher Mitarbeiter  
Chair of Methodology and Empirical Social Research  
Otto-von-Guericke-Universität  

[jonas.volle@ovgu.de](mailto:jonas.volle@ovgu.de)

**Sprechstunde**: individuell nach vorheriger Anmeldung per [Mail](mailto:jonas.volle@ovgu.de)

Donnerstag, 06.06.2023

**Quelle:** Ich orientiere mich für diese Sitzung an den Kapiteln 3 und 6 aus dem Buch:  

McLevey, John. 2021. Doing Computational Social Science: A Practical Introduction. 1st ed. Thousand Oaks: SAGE Publications.


## Inhalt

- Datenmanagement mit pandas


## 1. Datenmanagement mit Pandas

Nun verwenden wir Python nicht mehr als allgemeine Programmiersprache, sondern zur Verarbeitung von Daten mit Hilfe spezieller Datenverwaltungs- und -analysepakete. Wir werden uns in erster Linie auf ein Paket namens *Pandas* stützen. Pandas wurde von Wes McKinney für die Analyse von Paneldaten entwickelt (daher der Name). Es verfügt über spezielle Datenstrukturen, Funktionen und Methoden, mit denen Sie die meisten Datenverarbeitungsvorgänge für strukturierte quantitative Daten erledigen können.

Pandas Dokumentation: https://pandas.pydata.org/docs/reference/index.html

### Import

In [1]:
import pandas as pd

### Datenimport mit pandas

Das Pandas-Paket macht es einfach, Daten aus einer externen Datei direkt in ein Dataframe-Objekt zu laden.

| Datentyp | Reader | Writer |
|----------|----------|----------|
| CSV    | `read_csv()`   | `to_csv()`   |
| JSON    | `read_json()`   | `to_json()`   |
| Stata   | `read_stata()`   | `to_stata()`   |
| SAS   | `read_sas()`   | NA   |
| SPSS   | `read_spss()`   | NA   |


In diesem Kapitel werden Daten aus dem VDEM-Datensatz (Varieties of Democracy) verwendet. VDEM ist ein laufendes Forschungsprojekt zur Messung des Niveaus der Demokratie in Regierungen auf der ganzen Welt, und es werden laufend aktualisierte Versionen des Datensatzes veröffentlicht. Die Forschung wird von einem Team aus mehr als 50 Sozialwissenschaftlern geleitet, die die Sammlung und Analyse von Experteneinschätzungen von mehr als 3200 Historikern und Länderexperten koordinieren. Auf der Grundlage dieser Einschätzungen hat das VDEM-Projekt eine bemerkenswert komplexe Reihe von Indikatoren entwickelt, die sich an fünf übergeordneten Facetten der Demokratie orientieren: Wahldemokratie, liberale Demokratie, partizipative Demokratie, deliberative Demokratie und egalitäre Demokratie. Der Datensatz reicht bis ins Jahr 1789 zurück und gilt als Goldstandard für quantitative Daten über globale demokratische Entwicklungen.  

Das Codebuch finden Sie hier: https://v-dem.net/documents/24/codebook_v13.pdf

In [2]:
pd.read_csv?

Signature:
pd.read_csv(
    filepath_or_buffer: 'FilePath | ReadCsvBuffer[bytes] | ReadCsvBuffer[str]',
    *,
    sep: 'str | None | lib.NoDefault' = <no_default>,
    delimiter: 'str | None | lib.NoDefault' = None,
    header: "int | Sequence[int] | None | Literal['infer']" = 'infer',
    names: 'Sequence[Hashable] | None | lib.NoDefault' = <no_default>,
    index_col: 'IndexLabel | Literal[False] | None' = None,
    usecols: 'UsecolsArgType' = None,
    dtype: 'DtypeArg | None' = None,
    engine: 'CSVEngine | None' = None,
    converters: 'Mapping[Hashable, Callable] | None' = None,
    true_values: 'list | None' = None,
    false_values: 'list | None' = None,
    skipinitialspace: 'bool' = False,
    skiprows: 'list[int] | int | Callable[[Hashable], bool] | None' = None,
    skipfooter: 'int' = 0,
    nrows: 'int | None' = None,
    na_values: 'Hashable | Iterable[Hashable] | Mapping[Hashable, Iterable[Hashable]] | None' = None,
    keep_default_na: 'bool' = True,
    na_filter: '

In [3]:
df = pd.read_csv('../data/vdem/V-Dem-CY-Full+Others-v10.csv',
                low_memory=False)

Wie viele Zeilen und Spalten? Das geht mit der `.shape` Methode.

In [4]:
df.shape

(27013, 48)

Welche Variablen benötigen wir? Wir wählen Spalten in einem DataFrame aus, indem wir den DataFrame aufrufen gefolt mit einer Liste in eckigen Klammern, die die Namen der Spalten enthält.

Wir können die Namen der Spalten (columns) mithilfe des Attributs `.columns` für den DataFrame ausdrucken:

In [5]:
list(df.columns)

['country_name',
 'country_text_id',
 'e_regiongeo',
 'year',
 'v2x_polyarchy',
 'v2x_libdem',
 'v2x_partipdem',
 'v2x_delibdem',
 'v2x_egaldem',
 'v2smprivex',
 'v2smpolsoc',
 'v2caviol',
 'e_boix_regime',
 'v2dlreason',
 'v2dlcommon',
 'v2dlcountr',
 'v2dlconslt',
 'v2dlengage',
 'v2dlencmps',
 'v2dlunivl',
 'v2cseeorgs',
 'v2csreprss',
 'v2cscnsult',
 'v2csprtcpt',
 'v2csgender',
 'v2csantimv',
 'v2csrlgrep',
 'v2csrlgcon',
 'v2mecenefm',
 'v2mecenefi',
 'v2mecenefibin',
 'v2mecrit',
 'v2merange',
 'v2mefemjrn',
 'v2meharjrn',
 'v2meslfcen',
 'v2mebias',
 'v2mecorrpt',
 'v2exrescon',
 'v2exbribe',
 'v2exembez',
 'v2excrptps',
 'v2exthftps',
 'v2cldiscm',
 'v2cldiscw',
 'v2clacfree',
 'v2clrelig',
 'v2clfmove']

In diesem Fall möchten wir die folgenden Variablen behalten: 

1. den Ländernamen 
2. die Länder-ID 
3. die geografische Region 
4. Die Länder-ID 
3. die geographische Region 
4. Das Jahr 5. Der Polyarchie-Index 
6. Der Index der liberalen Demokratie
7. Der Index der partizipativen Demokratie 
8. Der Index der deliberativen Demokratie 
9. Der Index der egalitären Demokratie 
10. Ob die Privatsphäre der Internetnutzer und ihre Daten rechtlich geschützt sind 
11. Wie polarisiert das Land in politischen Fragen ist 
12. Ausmaß der politischen Gewalt 
13. Ob das Land eine Demokratie ist oder nicht 


In [6]:
subset_vars = ['country_name', 'country_text_id', 'e_regiongeo', 
               'year', 'v2x_polyarchy', 'v2x_libdem', 'v2x_partipdem', 
               'v2x_delibdem', 'v2x_egaldem', 'v2smprivex', 'v2smpolsoc', 
               'v2caviol', 'e_boix_regime']

sdf = df[subset_vars]

In [7]:
sdf.shape

(27013, 13)

In [8]:
list(sdf.columns)

['country_name',
 'country_text_id',
 'e_regiongeo',
 'year',
 'v2x_polyarchy',
 'v2x_libdem',
 'v2x_partipdem',
 'v2x_delibdem',
 'v2x_egaldem',
 'v2smprivex',
 'v2smpolsoc',
 'v2caviol',
 'e_boix_regime']

### Was ist im DataFrame?

Mit der Methode `.info()` können wir die Gesamtzahl der Beobachtungen, die Gesamtzahl der Spalten, die Namen der Spalten, die Anzahl der nicht fehlenden Beobachtungen für jede Variable, den Datentyp für jede Variable, die Anzahl der Variablen, die Daten jedes Typs enthalten (z. B. Ganzzahlen und Fließkommazahlen), und die Gesamtmenge des vom DataFrame verwendeten Speichers anzeigen:

In [9]:
sdf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 27013 entries, 0 to 27012
Data columns (total 13 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   country_name     27013 non-null  object 
 1   country_text_id  27013 non-null  object 
 2   e_regiongeo      27013 non-null  int64  
 3   year             27013 non-null  int64  
 4   v2x_polyarchy    25342 non-null  float64
 5   v2x_libdem       24350 non-null  float64
 6   v2x_partipdem    24923 non-null  float64
 7   v2x_delibdem     18557 non-null  float64
 8   v2x_egaldem      18557 non-null  float64
 9   v2smprivex       3562 non-null   float64
 10  v2smpolsoc       3562 non-null   float64
 11  v2caviol         12745 non-null  float64
 12  e_boix_regime    16270 non-null  float64
dtypes: float64(9), int64(2), object(2)
memory usage: 2.7+ MB


Die Datentypen in diesem Datenrahmen sind float64 (Zahlen mit Nachkommastellen), int64 (Ganzzahlen) und object. In Pandas bezieht sich object auf Spalten, die Strings oder gemischte Typen wie Strings und Integers enthalten (object umfasst auch viele andere Dinge: es ist eine Sammelkategorie). Pandas kann auch mit Booleans (Wahr oder Falsch), kategorischen Variablen und einigen speziellen Datetime-Objekten arbeiten. Erinnern Sie sich daran, wie wir die Spalten für unser Dataset ausgewählt haben. Im folgenden Code verwende ich dieselbe Idee, um nur einige wenige Variablen anzuzeigen. Wir werden dies später in diesem Kapitel noch etwas genauer erklären.

Wir können auch die Methode `.describe()` verwenden, um zusammenfassende Informationen über die quantitativen Variablen in unserem Datensatz zu erhalten, einschließlich der Anzahl der nicht fehlenden Informationen, des Mittelwerts und der Standardabweichung.


In [10]:
sdf[['e_regiongeo', 'year', 'v2x_polyarchy']].describe()

e_regiongeo          year  v2x_polyarchy
count  27013.000000  27013.000000   25342.000000
mean       9.266575   1926.556177       0.262821
std        5.733007     63.754335       0.260966
min        1.000000   1789.000000       0.007000
25%        4.000000   1878.000000       0.057000
50%        8.000000   1937.000000       0.171500
75%       14.000000   1980.000000       0.367000
max       19.000000   2019.000000       0.924000

### Heads, tails, and samples

Wir können auch den "Kopf" oder das "Ende" unseres Datenrahmens mit den Methoden `.head() ` und `.tail()` untersuchen, die standardmäßig die ersten oder letzten fünf Zeilen in einem DataFrame verwenden, es sei denn, Sie geben eine andere Zahl als Argument an, wie z. B. `.head(10)`:

In [11]:
sdf[['country_name', 'year', 'v2x_libdem']].head()

country_name  year  v2x_libdem
0       Mexico  1789       0.043
1       Mexico  1790       0.040
2       Mexico  1791       0.040
3       Mexico  1792       0.040
4       Mexico  1793       0.040

In [12]:
sdf[['country_name', 'year', 'v2x_libdem']].tail(3)

country_name  year  v2x_libdem
27010  Piedmont-Sardinia  1859       0.150
27011  Piedmont-Sardinia  1860       0.148
27012  Piedmont-Sardinia  1861       0.149

Wenn Sie eine Zufallsstichprobe von Zeilen bevorzugen, können Sie die Methode `.sample()` verwenden, bei der Sie die Anzahl der Zeilen angeben müssen, die Sie abfragen möchten:

In [13]:
sdf[['country_name', 'year', 'v2x_libdem']].sample(15)

country_name  year  v2x_libdem
17946            Namibia  1967       0.051
23263             Norway  1943       0.087
5210         Afghanistan  1824       0.022
6462         South Korea  1957       0.158
24395             Serbia  1841       0.060
951         South Africa  1929       0.125
4611                Peru  1950       0.038
19034         Uzbekistan  1854         NaN
7328            Thailand  2017       0.106
25637              Baden  1803         NaN
4071               Haiti  2016       0.243
6463         South Korea  1958       0.150
20808  Equatorial Guinea  1940       0.018
8945               Niger  1996       0.210
16058              China  1945       0.056

### Zeilen filtern

Bei der Ausführung der Methode `.describe()` haben Sie vielleicht bemerkt, dass der Bereich für die Jahresvariable 1789-2019 ist. Angenommen, wir haben einen guten Grund, uns auf die Jahre von 1900 bis 2019 zu konzentrieren. Dann müssen wir die Daten filtern, um nur die Zeilen zu erhalten, die unseren Anforderungen entsprechen. Es gibt mehrere Möglichkeiten, Zeilen zu filtern, einschließlich Slices (z. B. alle Beobachtungen zwischen Index i und Index j) oder nach einer expliziten Bedingung, wie z. B. "Zeilen, in denen das Jahr >= 1900". Beachten Sie, dass, wenn wir einen DataFrame filtern oder zerschneiden, das neue Objekt nur eine Ansicht des Originals ist und sich immer noch auf die gleichen Daten bezieht. Pandas warnt uns, wenn wir versuchen, das gefilterte Objekt zu verändern, so dass es in den meisten Fällen einfacher ist, eine neue Kopie zu erstellen.

In [14]:
fsdf = sdf[sdf['year'] >= 1900].copy()
fsdf.info()

<class 'pandas.core.frame.DataFrame'>
Index: 18787 entries, 111 to 25622
Data columns (total 13 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   country_name     18787 non-null  object 
 1   country_text_id  18787 non-null  object 
 2   e_regiongeo      18787 non-null  int64  
 3   year             18787 non-null  int64  
 4   v2x_polyarchy    18663 non-null  float64
 5   v2x_libdem       18424 non-null  float64
 6   v2x_partipdem    18587 non-null  float64
 7   v2x_delibdem     18557 non-null  float64
 8   v2x_egaldem      18557 non-null  float64
 9   v2smprivex       3562 non-null   float64
 10  v2smpolsoc       3562 non-null   float64
 11  v2caviol         12745 non-null  float64
 12  e_boix_regime    11841 non-null  float64
dtypes: float64(9), int64(2), object(2)
memory usage: 2.0+ MB


In [15]:
sdf[(sdf['year'] >= 1900) & (sdf['country_name'] == "Mexico")].head()

country_name country_text_id  e_regiongeo  year  v2x_polyarchy  \
111       Mexico             MEX           17  1900          0.118   
112       Mexico             MEX           17  1901          0.118   
113       Mexico             MEX           17  1902          0.118   
114       Mexico             MEX           17  1903          0.118   
115       Mexico             MEX           17  1904          0.118   

     v2x_libdem  v2x_partipdem  v2x_delibdem  v2x_egaldem  v2smprivex  \
111       0.061          0.098         0.034        0.028         NaN   
112       0.061          0.098         0.034        0.028         NaN   
113       0.061          0.098         0.034        0.028         NaN   
114       0.061          0.098         0.034        0.028         NaN   
115       0.061          0.098         0.034        0.028         NaN   

     v2smpolsoc  v2caviol  e_boix_regime  
111         NaN     1.323            0.0  
112         NaN     1.323            0.0  
113         NaN     1.323            0.0  
114         NaN     1.323            0.0  
115         NaN     1.323            0.0

### Schreiben von Dateien

So wie wir unsere ursprüngliche CSV-Datei mit der Funktion `read_csv()` in Pandas eingelesen habe, können wir diesen neuen Dataframe mit der Methode `to_csv()` auch wieder schreiben:

In [16]:
fsdf.to_csv('../data/vdem/filtered_subset.csv', index=False)

<div class='alert alert-block alert-success'>

### Aufgabe 1

--> session_04_excercise_1.ipynb

</div>

## Pandas Datenstruktur

### The series

Jede Spalte in einem dataframe ist ein Objekt namens Series. Eine Series ist ein eindimensionales Objekt (z. B. ein Zahlenvektor) mit einem Index, der selbst ein Vektor oder ein Array von Bezeichnungen ist.

Zum Beispiel ist die Spalte *v2x_delibdem* in fsdf eine Series, die Floats und die Indexbezeichnung für jede Beobachtung enthält. Wenn ich eine Stichprobe von 15 Beobachtungen ausdrucke, erhalte ich einen numerischen Index für jede Beobachtung auf der linken Seite und den tatsächlichen Wert auf der rechten Seite. Die Indexwerte sind in der Series selbst geordnet, aber hier sind sie nicht in der richtigen Reihenfolge, weil wir eine Zufallsstichprobe gezogen haben. Da dies nur zu Demonstrationszwecken dient, habe ich einen random_state-Wert eingefügt, um sicherzustellen, dass Sie die gleiche Stichprobe erhalten wie ich, wenn Sie diesen Block erneut ausführen.

In [17]:
fsdf['v2x_delibdem']

111      0.034
112      0.034
113      0.034
114      0.034
115      0.034
         ...  
25618    0.215
25619    0.194
25620    0.168
25621    0.213
25622    0.183
Name: v2x_delibdem, Length: 18787, dtype: float64

In [18]:
fsdf['v2x_delibdem'].sample(15, random_state = 42)

8437     0.305
25197    0.059
6661     0.079
19266    0.390
7633     0.632
1230     0.727
8829     0.020
10118    0.008
8996     0.047
2360     0.048
18407    0.018
153      0.132
10794    0.805
14375    0.800
24240      NaN
Name: v2x_delibdem, dtype: float64

In den meisten Fällen ist der Standardindex für eine Series oder einen DataFrame ein unveränderlicher Vektor aus ganzen Zahlen:

In [19]:
fsdf.index

Index([  111,   112,   113,   114,   115,   116,   117,   118,   119,   120,
       ...
       25613, 25614, 25615, 25616, 25617, 25618, 25619, 25620, 25621, 25622],
      dtype='int64', length=18787)

Wir können einen Index leicht so ändern, dass er stattdessen aus einem anderen Vektortyp besteht, z. B. einem string. Überraschenderweise müssen Indexwerte nicht eindeutig sein. Dies ermöglicht einige leistungsfähige Techniken, aber die meiste Zeit sollten Sie vermeiden, Indizes manuell zu ändern.

### Zugriff auf eine bestimmte Zeile über ihren Index

Wir können den Index verwenden, um bestimmte Zeilen aus einem dataframe oder bestimmte Werte aus einer series abzurufen, ganz so, als ob wir ein Element aus einer Liste, einem Tupel oder einem Array auswählen würden. Am einfachsten geht das, indem man den Indexwert (z. B. 202) an `.loc[]` übergibt. Wie Sie unten sehen können, ist das Ergebnis der beobachtungsspezifische Wert für jede Variable im dataframe:

In [20]:
fsdf.loc[202]

country_name       Mexico
country_text_id       MEX
e_regiongeo            17
year                 1991
v2x_polyarchy       0.435
v2x_libdem          0.221
v2x_partipdem       0.246
v2x_delibdem         0.31
v2x_egaldem         0.214
v2smprivex            NaN
v2smpolsoc            NaN
v2caviol            -0.71
e_boix_regime         0.0
Name: 202, dtype: object

Beachten Sie, dass sich `.loc` nicht auf die 202. Zeile des dataframe bezieht. Wenn Sie sich den obigen `.index`-Befehl genau angesehen haben, ist Ihnen vielleicht aufgefallen, dass der dataframe nur 18.787 Zeilen enthält, aber `.loc` immer noch Zeile 20.000 zurückgeben kann - der Index hat sich nicht geändert, als Sie eine Reihe von Zeilen aus dem dataframe entfernt haben. Stellen Sie sich `.loc` als Zugriff auf ein Wörterbuch der Indexwerte vor - es gibt sogar einen KeyError, wenn Sie nach einem Element fragen, das nicht existiert.

Wenn wir stattdessen auf die n-te Zeile eines dataframe zugreifen wollen, können wir `.iloc[n]` verwenden. Stellen Sie sich den Index als eine Liste vor, und Sie beziehen sich auf ein Element dieser Liste durch seinen Listenindex. Verwenden wir `.iloc`, um das letzte Element im dataframe auszuwählen. Beachten Sie, dass die Indexposition für das letzte Element 18.786 sein wird, obwohl die Länge des dataframe 18.787 beträgt, da Python-Datenstrukturen fast immer mit einem Null-Index versehen sind. Hier sehen Sie den Index der Zeile, der früher die Zeilennummer war, als Name am unteren Rand:

In [21]:
len(fsdf)

18787

In [22]:
fsdf.iloc[18786]

country_name       Zanzibar
country_text_id         ZZB
e_regiongeo               8
year                   2019
v2x_polyarchy         0.245
v2x_libdem             0.18
v2x_partipdem         0.083
v2x_delibdem          0.183
v2x_egaldem           0.187
v2smprivex           -1.843
v2smpolsoc           -0.848
v2caviol             -0.181
e_boix_regime           NaN
Name: 25622, dtype: object

In [23]:
fsdf.tail()

country_name country_text_id  e_regiongeo  year  v2x_polyarchy  \
25618     Zanzibar             ZZB            8  2015          0.316   
25619     Zanzibar             ZZB            8  2016          0.267   
25620     Zanzibar             ZZB            8  2017          0.260   
25621     Zanzibar             ZZB            8  2018          0.262   
25622     Zanzibar             ZZB            8  2019          0.245   

       v2x_libdem  v2x_partipdem  v2x_delibdem  v2x_egaldem  v2smprivex  \
25618       0.240          0.114         0.215        0.255      -1.749   
25619       0.212          0.090         0.194        0.234      -1.749   
25620       0.171          0.097         0.168        0.213      -1.749   
25621       0.170          0.095         0.213        0.214      -1.749   
25622       0.180          0.083         0.183        0.187      -1.843   

       v2smpolsoc  v2caviol  e_boix_regime  
25618      -0.210    -0.321            NaN  
25619      -1.512    -0.181            NaN  
25620      -1.512    -0.181            NaN  
25621      -1.389    -0.181            NaN  
25622      -0.848    -0.181            NaN

Wenn es keinen Grund gibt, die ursprüngliche Indizierung des ungefilterten dataframe beizubehalten, ist es in der Regel eine gute Idee, den Index zurückzusetzen:

In [24]:
fsdf.reset_index(drop = True) 
fsdf.loc[18786]

country_name       Turkmenistan
country_text_id             TKM
e_regiongeo                  11
year                       2004
v2x_polyarchy             0.146
v2x_libdem                0.027
v2x_partipdem              0.06
v2x_delibdem              0.017
v2x_egaldem                 0.1
v2smprivex               -1.395
v2smpolsoc                1.826
v2caviol                 -2.387
e_boix_regime               0.0
Name: 18786, dtype: object

Danach werden `.loc` und `.iloc` ziemlich austauschbar, mit ein paar Ausnahmen: `.loc` hat wörterbuchähnliche Fähigkeiten, während `.iloc` eher listenähnlich ist. Schauen wir uns nun den dataframe genauer an.

### Dataframes

Dataframes in Pandas sind eigentlich nur Sammlungen von Series, die an denselben Indexwerten ausgerichtet sind. Mit anderen Worten: Die Serien, mit denen wir zuvor gearbeitet haben, haben ihre eigenen Indizes, wenn wir mit ihnen als eigenständige Serien arbeiten, aber im fsdf dataframe teilen sie sich einen Index.  
Wie Sie bereits gesehen haben, sind dataframes mit Variablen in den Spalten und Beobachtungen in den Zeilen organisiert, und Sie können eine einzelne series aus einem dataframe mit eckigen Klammern holen - lassen Sie uns das jetzt mit dem fsdf dataframe tun:

In [25]:
deliberative = fsdf['v2x_delibdem']

Beachten Sie, dass wir auch die Punktnotation verwenden können, um Spalten auszuwählen. `fsdf.v2x_delibdem` ist funktional gleichwertig mit `fsdf['v2x_delibdem']` und kann austauschbar verwendet werden.

In [26]:
fsdf.v2x_delibdem

111      0.034
112      0.034
113      0.034
114      0.034
115      0.034
         ...  
25618    0.215
25619    0.194
25620    0.168
25621    0.213
25622    0.183
Name: v2x_delibdem, Length: 18787, dtype: float64

Wir sind nicht darauf beschränkt, Spalten auszuwählen, die bereits in unserem Datensatz vorhanden sind. Sie können auch neue Spalten erstellen und hinzufügen. Sie können beispielsweise eine neue Spalte mit der Bezeichnung "21. Jahrhundert" erstellen und einen booleschen Wert zuweisen, der darauf basiert, ob die Beobachtung in den 2000er Jahren liegt:

In [27]:
fsdf['21 Century'] = fsdf['year'] >= 2000 
fsdf[['21 Century']].value_counts()

21 Century
False         15225
True           3562
Name: count, dtype: int64

Manchmal handelt es sich bei den neu erstellten Spalten um Umwandlungen einer bereits im dataframe vorhandenen series. Sie können beispielsweise eine neue Spalte "missing_political_violence_data" erstellen, die den Wert "True" annimmt, wenn die Serie "v2caviol" (Ausmaß der politischen Gewalt) leer ist, und andernfalls "False":

In [28]:
fsdf['missing_political_violence_data'] = fsdf['v2caviol'].isna()
fsdf['missing_political_violence_data'].value_counts()

missing_political_violence_data
False    12745
True      6042
Name: count, dtype: int64

Wie Sie bei der Ausführung von value_counts() sehen können, fehlen für 6042 Beobachtungen Daten zum Ausmaß der politischen Gewalt.

### Missing data

Es ist wichtig zu verstehen, wie mit fehlenden Daten umgegangen wird. Fehlende Daten kommen in realen Datensätzen häufig vor, und sie können aus mehreren Gründen fehlen! Im Allgemeinen verwendet Pandas den Wert np.nan, um fehlende Daten darzustellen. Der np.nan-Wert von NumPy ist ein Spezialfall einer Fließkommazahl, die einen nicht darstellbaren Wert darstellt. Diese Art von Werten werden NaNs (Not a Number) genannt.

In [29]:
import numpy as np 
np.nan

nan

np.nan kann nicht in Gleichheitstests verwendet werden, da jeder Vergleich mit einem np.nan-Wert als Falsch ausgewertet wird. Dies schließt den Vergleich von np.nan mit sich selbst ein.  

np.nan-Werte werden nicht zu False oder None ausgewertet. Dies kann es schwierig machen, fehlende Werte zu unterscheiden. Zu diesem Zweck können Sie die Funktion np.isnan() verwenden, die besonders im Kontrollfluss nützlich ist.

In [30]:
np.nan == None

False

In [31]:
np.nan == np.nan

False

In [32]:
np.isnan(np.nan)

True

In [33]:
fsdf[['country_name', 'year', 'v2caviol']][np.isnan(fsdf['v2caviol'])]

country_name  year  v2caviol
231       Suriname  1900       NaN
232       Suriname  1901       NaN
233       Suriname  1902       NaN
234       Suriname  1903       NaN
235       Suriname  1904       NaN
...            ...   ...       ...
25352      Hungary  1913       NaN
25353      Hungary  1914       NaN
25354      Hungary  1915       NaN
25355      Hungary  1916       NaN
25356      Hungary  1917       NaN

[6042 rows x 3 columns]

In [34]:
fsdf[['country_name', 'year', 'v2caviol']][fsdf['v2caviol'].isna()]

country_name  year  v2caviol
231       Suriname  1900       NaN
232       Suriname  1901       NaN
233       Suriname  1902       NaN
234       Suriname  1903       NaN
235       Suriname  1904       NaN
...            ...   ...       ...
25352      Hungary  1913       NaN
25353      Hungary  1914       NaN
25354      Hungary  1915       NaN
25355      Hungary  1916       NaN
25356      Hungary  1917       NaN

[6042 rows x 3 columns]

Außerdem werden np.nan-Werte im Allgemeinen von Pandas-Funktionen ausgeschlossen, die Berechnungen über dataframes, Zeilen oder Spalten durchführen. In der Dokumentation wird zum Beispiel oft angegeben, dass eine Berechnung über alle Werte durchgeführt wird, wobei NaN- oder NULL-Werte ausgeschlossen werden.

In [35]:
total = len(fsdf['v2caviol']) 
count = fsdf['v2caviol'].count() 
print(f'Total: {total}') 
print(f'Count: {count}') 
print(f'Diff: {total-count}')

Total: 18787
Count: 12745
Diff: 6042


Die Gesamtzahl der Elemente in der Spalte v2caviol (politische Gewalt) ist viel höher als die von der Funktion `count()` erhaltene Anzahl. Wenn das, was wir oben gelernt haben, richtig ist, sollte dieser Unterschied berücksichtigt werden, wenn wir herausfinden, wie viele Elemente in dieser Spalte NaNs sind.

In [36]:
nans = fsdf['v2caviol'].isna().sum()
print(' NaNs: {}'.format(nans))

 NaNs: 6042


Wie Sie wahrscheinlich wissen, kann die Methode `.isna()`, die der Methode `np.isnan()` ähnelt, aber zusätzliche Fälle abdeckt, beim Transformieren und Filtern von Daten sehr nützlich sein.

### Aggregieren und Gruppieren

Datenanalyseprojekte beinhalten oft Aggregations- oder Gruppierungsoperationen. So kann es beispielsweise erforderlich sein, zusammenfassende Statistiken für Beobachtungen zu berechnen und zu vergleichen, die unterschiedliche Werte für eine kategoriale Variable annehmen. Es kann hilfreich sein, den Datensatz selbst aufzuschlüsseln und Operationen auf verschiedenen Teilmengen von Daten durchzuführen. Dazu verwenden wir die Methode `.groupby()`, die den dataframe auf der Grundlage der Werte einer bestimmten Variable in Gruppen unterteilt. Anschließend können wir Operationen mit den resultierenden Gruppen durchführen. Wir gruppieren unsere Länder in geografische Regionen, indem wir die Variable *e_regiongeo* verwenden:

In [37]:
grouped = fsdf.groupby('e_regiongeo')

Der obige Code gibt ein gruppiertes Objekt zurück, mit dem wir arbeiten können. Nehmen wir an, wir wollen eine bestimmte Gruppe herausziehen, z. B. Südostasien, die in den Daten mit der numerischen ID 13 dargestellt wird. Ich weiß das, weil die entsprechenden Informationen im VDEM-Codebuch zu finden sind, das Sie bei Ihrer Arbeit mit den VDEM-Daten immer geöffnet halten sollten.

Wir können die Methode get_group() verwenden, um eine Gruppe aus dem gruppierten Objekt zu ziehen. (Beachten Sie, dass der nachstehende Code .get_group() fsdf[fsdf['e_regiongeo'] == 13] entspricht).

In [38]:
south_east_asia = grouped.get_group(13)

In [39]:
south_east_asia[['country_name', 'year', 'v2x_polyarchy']].head()

country_name  year  v2x_polyarchy
1384  Burma/Myanmar  1900          0.091
1385  Burma/Myanmar  1901          0.091
1386  Burma/Myanmar  1902          0.091
1387  Burma/Myanmar  1903          0.091
1388  Burma/Myanmar  1904          0.091

Bei den in south_east_asia gespeicherten Daten handelt es sich um alle Beobachtungen der südostasiatischen Länder in den VDEM-Daten, die nun in einem eigenen dataframe gespeichert sind. .get_group() ist eine weitere Möglichkeit, eine Teilmenge eines dataframes zu extrahieren (mit Hilfe eines groupby-Objekts), und ist besonders nützlich, wenn die Teilmenge der Daten, mit der Sie arbeiten möchten, nur Beobachtungen mit einem bestimmten Wert für eine kategoriale Variable in Ihren Daten sind. Wenn wir einen Datensatz auf diese Weise gruppieren, geschieht dies im Allgemeinen, weil wir etwas für eine Gruppe innerhalb des Datensatzes oder für mehrere Gruppen, die wir vergleichen möchten, berechnen möchten. Dazu geben wir das gruppierte Objekt, die Serie, für die wir eine Operation durchführen möchten, und schließlich die Operation, die wir durchführen möchten, an. Berechnen wir zum Beispiel den Median des Polyarchie-Scores für Länder in jeder der Regionen im Datensatz:

In [40]:
poly = grouped['v2x_polyarchy'].median() 
poly.head()

e_regiongeo
1    0.7940
2    0.8170
3    0.3890
4    0.2385
5    0.0955
Name: v2x_polyarchy, dtype: float64

Es wäre nützlicher, den Namen der Region zu sehen als ihre numerische Bezeichnung. Wir können dies erreichen, indem wir ein Wörterbuch erstellen, das die numerischen IDs auf den Regionsnamen abbildet, und dann die Methode .map() verwenden, um Pandas mitzuteilen, wo es die Werte nachschlagen soll, die es benötigt, um eine neue Spalte mit den Ländernamen zu erstellen. Zuerst das Wörterbuch:

In [41]:
regions = { 1:'Western Europe', 
           2:'Northern Europe', 
           3:'Southern Europe', 
           4:'Eastern Europe', 
           5:'Northern Africa',
           6:'Western Africa', 
           7:'Middle Africa', 
           8:'Eastern Africa', 
           9:'Southern Africa', 
           10:'Western Asia', 
           11:'Central Asia', 
           12:'East Asia', 
           13:'South-East Asia', 
           14:'South Asia', 
           15:'Oceania', # (including Australia and the Pacific) 
           16:'North America', 
           17:'Central America', 
           18:'South America', 
           19:'Caribbean' # (including Belize Cuba Haiti Dominican Republic) 
          }

Nun können wir dieses Wörterbuch an die Methode `.map()` übergeben, die auf die Serie `fsdf['e_regiongeo']` angewendet wird, und so eine neue Serie namens `fsdf['Region']` erstellen:

In [42]:
fsdf['Region'] = fsdf['e_regiongeo'].map(regions)

Es ist auch möglich, nach mehreren Variablen zu gruppieren, z. B. nach geografischer Region und Jahr, und dann eine Operation an diesen etwas feineren Gruppen durchzuführen. Auf diese Weise ergeben sich 2211 Gruppen, so dass wir eine Zufallsstichprobe von 10 Gruppen ansehen werden:

In [43]:
grouped = fsdf.groupby(['Region', 'year']) 
poly = grouped['v2x_polyarchy'].median() 
poly.reset_index()
pd.DataFrame(poly).reset_index()

Region  year  v2x_polyarchy
0          Caribbean  1900         0.1340
1          Caribbean  1901         0.1690
2          Caribbean  1902         0.1565
3          Caribbean  1903         0.1525
4          Caribbean  1904         0.1700
...              ...   ...            ...
2206  Western Europe  2015         0.8830
2207  Western Europe  2016         0.8830
2208  Western Europe  2017         0.8770
2209  Western Europe  2018         0.8730
2210  Western Europe  2019         0.8730

[2211 rows x 3 columns]

Für das gruppierte Objekt selbst können andere Operationen durchgeführt werden, z. B. die Berechnung der Anzahl der Beobachtungen in jeder Gruppe (entspricht value_counts()):

In [44]:
grouped.size().sort_values(ascending=False)

Region          year
Eastern Africa  2015    20
                2012    20
                2019    20
                2018    20
                2017    20
                        ..
Central Asia    1904     1
                1903     1
                1902     1
                1901     1
                1914     1
Length: 2211, dtype: int64

In [45]:
grouped.size().sort_values(ascending=False).reset_index(name='count')

Region  year  count
0     Eastern Africa  2015     20
1     Eastern Africa  2012     20
2     Eastern Africa  2019     20
3     Eastern Africa  2018     20
4     Eastern Africa  2017     20
...              ...   ...    ...
2206    Central Asia  1904      1
2207    Central Asia  1903      1
2208    Central Asia  1902      1
2209    Central Asia  1901      1
2210    Central Asia  1914      1

[2211 rows x 3 columns]

Schließlich können wir mit der Methode `agg()` mehrere Operationen auf ein gruppiertes Objekt anwenden. Die `agg()`-Methode wendet eine oder mehrere Aggregatfunktionen auf ein gruppiertes Objekt an und gibt die Ergebnisse der einzelnen Funktionen zurück:

In [46]:
with_agg = grouped['v2x_polyarchy'].agg(['min', 'median', 'max', 'count']) 
with_agg.reset_index()

Region  year    min  median    max  count
0          Caribbean  1900  0.080  0.1340  0.222      5
1          Caribbean  1901  0.080  0.1690  0.222      6
2          Caribbean  1902  0.080  0.1565  0.295      6
3          Caribbean  1903  0.080  0.1525  0.315      6
4          Caribbean  1904  0.080  0.1700  0.325      6
...              ...   ...    ...     ...    ...    ...
2206  Western Europe  2015  0.867  0.8830  0.904      7
2207  Western Europe  2016  0.852  0.8830  0.904      7
2208  Western Europe  2017  0.835  0.8770  0.896      7
2209  Western Europe  2018  0.806  0.8730  0.893      7
2210  Western Europe  2019  0.809  0.8730  0.882      7

[2211 rows x 6 columns]

Wir können sogar unsere eigene Funktion definieren, die agg() verwenden soll! Wenn wir bereit sind, ein Wörterbuch zu übergeben, können wir mit .agg() auch verschiedene Funktionen auf mehrere Variablen gleichzeitig anwenden! Anstatt eine Liste pro Funktion zu übergeben, können Sie ein Wörterbuch verwenden, in dem die Spaltennamen die Schlüssel und die Funktionen die Werte sind (Sie können auch eine Liste von Funktionen übergeben), um eine wirklich komplexe Aggregation in einer einzigen Codezeile durchzuführen.

## Kombinierung von DataFrames

Das Kombinieren von dataframes ist eine sehr häufige Aufgabe. Auch wenn es nicht offensichtlich erscheint, ist das Kombinieren von Datensätzen eine der wertvollsten Fähigkeiten, die man in der rechnergestützten Sozialwissenschaft haben kann. Im Folgenden werden einige der gebräuchlichsten Ansätze betrachtet, nämlich die Verkettung und Zusammenführung, und es wird kurz eine fortgeschrittenere Reihe von Methoden beschrieben, die gemeinhin als Datensatzverknüpfung bezeichnet werden.  

Die Verkettung eines dataframes ist konzeptionell recht einfach - stellen Sie sich vor, Sie hängen die Zeilen oder Spalten eines dataframes unter oder rechts von der letzten Zeile oder Spalte eines anderen dataframes an. Damit dies sinnvoll ist, sollten die beiden dataframes mindestens eine Zeile oder Spalte gemeinsam haben.

In [47]:
full_df = pd.read_csv("../data/vdem/filtered_subset.csv") 
df_australia = full_df[full_df.country_name == "Australia"].copy()
len(df_australia)

120

In [48]:
df_australia.head()

country_name country_text_id  e_regiongeo  year  v2x_polyarchy  \
6773    Australia             AUS           15  1900          0.641   
6774    Australia             AUS           15  1901          0.651   
6775    Australia             AUS           15  1902          0.789   
6776    Australia             AUS           15  1903          0.789   
6777    Australia             AUS           15  1904          0.804   

      v2x_libdem  v2x_partipdem  v2x_delibdem  v2x_egaldem  v2smprivex  \
6773         NaN          0.427           NaN          NaN         NaN   
6774       0.566          0.463         0.556        0.427         NaN   
6775       0.706          0.568         0.689        0.551         NaN   
6776       0.705          0.569         0.688        0.550         NaN   
6777       0.705          0.574         0.699        0.562         NaN   

      v2smpolsoc  v2caviol  e_boix_regime  
6773         NaN       NaN            NaN  
6774         NaN       NaN            1.0  
6775         NaN       NaN            1.0  
6776         NaN       NaN            1.0  
6777         NaN       NaN            1.0

In [49]:
df_sa = full_df[full_df.country_name == "South Africa"].copy()
len(df_sa)

120

In [50]:
df_sa.tail()

country_name country_text_id  e_regiongeo  year  v2x_polyarchy  \
713  South Africa             ZAF            9  2015          0.713   
714  South Africa             ZAF            9  2016          0.715   
715  South Africa             ZAF            9  2017          0.712   
716  South Africa             ZAF            9  2018          0.681   
717  South Africa             ZAF            9  2019          0.678   

     v2x_libdem  v2x_partipdem  v2x_delibdem  v2x_egaldem  v2smprivex  \
713       0.602          0.451         0.608        0.453       1.464   
714       0.599          0.452         0.623        0.434       1.464   
715       0.604          0.452         0.621        0.497       1.464   
716       0.585          0.442         0.568        0.449       1.464   
717       0.575          0.434         0.586        0.455       0.253   

     v2smpolsoc  v2caviol  e_boix_regime  
713      -0.232     0.598            NaN  
714      -0.232     0.598            NaN  
715      -0.232     0.625            NaN  
716      -0.232     0.625            NaN  
717      -0.404     0.625            NaN

In [51]:
df_sa[df_sa.country_name.isna()]

Empty DataFrame
Columns: [country_name, country_text_id, e_regiongeo, year, v2x_polyarchy, v2x_libdem, v2x_partipdem, v2x_delibdem, v2x_egaldem, v2smprivex, v2smpolsoc, v2caviol, e_boix_regime]
Index: []

Standardmäßig führt pd.concat() eine zeilenweise Verknüpfung durch, die als Achse=0 bezeichnet wird. Diese Vorgabe kann durch die Angabe von Achse=1 außer Kraft gesetzt werden, wodurch eine spaltenweise Verknüpfung erfolgt:

In [52]:
concatenated = pd.concat([df_australia, df_sa], axis=0)
len(concatenated)

240

In [53]:
concatenated.head()

country_name country_text_id  e_regiongeo  year  v2x_polyarchy  \
6773    Australia             AUS           15  1900          0.641   
6774    Australia             AUS           15  1901          0.651   
6775    Australia             AUS           15  1902          0.789   
6776    Australia             AUS           15  1903          0.789   
6777    Australia             AUS           15  1904          0.804   

      v2x_libdem  v2x_partipdem  v2x_delibdem  v2x_egaldem  v2smprivex  \
6773         NaN          0.427           NaN          NaN         NaN   
6774       0.566          0.463         0.556        0.427         NaN   
6775       0.706          0.568         0.689        0.551         NaN   
6776       0.705          0.569         0.688        0.550         NaN   
6777       0.705          0.574         0.699        0.562         NaN   

      v2smpolsoc  v2caviol  e_boix_regime  
6773         NaN       NaN            NaN  
6774         NaN       NaN            1.0  
6775         NaN       NaN            1.0  
6776         NaN       NaN            1.0  
6777         NaN       NaN            1.0

In [54]:
concatenated.tail()

country_name country_text_id  e_regiongeo  year  v2x_polyarchy  \
713  South Africa             ZAF            9  2015          0.713   
714  South Africa             ZAF            9  2016          0.715   
715  South Africa             ZAF            9  2017          0.712   
716  South Africa             ZAF            9  2018          0.681   
717  South Africa             ZAF            9  2019          0.678   

     v2x_libdem  v2x_partipdem  v2x_delibdem  v2x_egaldem  v2smprivex  \
713       0.602          0.451         0.608        0.453       1.464   
714       0.599          0.452         0.623        0.434       1.464   
715       0.604          0.452         0.621        0.497       1.464   
716       0.585          0.442         0.568        0.449       1.464   
717       0.575          0.434         0.586        0.455       0.253   

     v2smpolsoc  v2caviol  e_boix_regime  
713      -0.232     0.598            NaN  
714      -0.232     0.598            NaN  
715      -0.232     0.625            NaN  
716      -0.232     0.625            NaN  
717      -0.404     0.625            NaN

Wenn wir die beiden dataframes verketten, bleibt die Anzahl der Spalten gleich, aber die Anzahl der Zeilen erhöht sich, da die Zeilen in beiden ursprünglichen dataframes berücksichtigt werden. Normalerweise würde diese Art der Verkettung zu einer unterschiedlichen Anzahl von Spalten führen, aber in diesem Fall hatten die beiden Dataframes, die wir zusammengefügt haben, genau die gleichen Spalten (was sinnvoll ist, da sie beide aus demselben übergeordneten Dataframe extrahiert wurden).

### Merging

Eine andere Möglichkeit, Datensätze zu kombinieren, ist ihre Zusammenführung. Wenn Sie einen dataframe erstellen möchten, der Spalten aus mehreren Datensätzen enthält, aber in den Zeilen nach einer bestimmten Spalte (oder einem Satz von Spalten) ausgerichtet ist, sollten Sie die Funktion merge() verwenden. Um dies zu veranschaulichen, werden wir mit Daten aus zwei verschiedenen Quellen arbeiten. Die erste Quelle sind die VDEM-Daten, die wir im ersten Teil dieses Kapitels verwendet haben (fsdf). Die zweite ist ein Datensatz von Freedom House über den Grad der Internetfreiheit in 65 Ländern. Weitere Informationen finden Sie unter https://freedomhouse.org/countries/freedom-net/scores.

In [55]:
freedom_df = pd.read_csv( "../data/freedom_house/internet_freedoms_2020.csv")

Um diese dataframes zusammenzuführen, müssen wir eine Spalte finden, mit der wir die Zeilen des einen dataframes mit den Zeilen des anderen abgleichen können. Die Spalten müssen nicht den gleichen Namen haben, sondern nur Werte, die einander zugeordnet werden können. Die von uns gewählten Spalten werden in unserer Zusammenführung als "Schlüssel" bezeichnet. In unserem Fall sind dies die Spalten mit den Ländernamen aus jedem Datensatz.

In [56]:
list(fsdf.columns)

['country_name',
 'country_text_id',
 'e_regiongeo',
 'year',
 'v2x_polyarchy',
 'v2x_libdem',
 'v2x_partipdem',
 'v2x_delibdem',
 'v2x_egaldem',
 'v2smprivex',
 'v2smpolsoc',
 'v2caviol',
 'e_boix_regime',
 '21 Century',
 'missing_political_violence_data',
 'Region']

In [57]:
list(freedom_df.columns)

['Country',
 'Total Score',
 'Status',
 'Obstacles to Access',
 'Limits on Content',
 'Violations of User Rights']

Wir verwenden die Funktion "Zusammenführen", um diese beiden dataframes mit "country_name" und "Country" zu kombinieren. Wir führen einen inner-join durch, was der Standard ist, wenn die Option nicht gesetzt ist, und behalten nur die Schlüssel (d. h. Länder), die in beiden dataframes erscheinen:

In [58]:
merged = pd.merge(fsdf, freedom_df, how='left', 
                  left_on='country_name', right_on='Country')

In [59]:
len(merged), len(merged.columns)

(18787, 22)

In [60]:
len(freedom_df), len(freedom_df.columns)

(65, 6)

In [61]:
len(fsdf), len(fsdf.columns)

(18787, 16)

Sie sollten fünf neue Spalten in dem zusammengeführten dataframe im Vergleich zu dem fsdf-Dataframe sehen. Beachten Sie, wie viele Zeilen jeder der beiden dataframes hat: viel weniger Zeilen als der ursprüngliche VDEM dataframe, aber viel mehr als der Freedom House dataframe. Wenn also in unserem Fall das Land einer Zeile in dem anderen Datensatz nicht vorkommt, wird diese Zeile nicht in den zusammengeführten dataframe aufgenommen.  

Dies kann mit dem Parameter how angepasst werden. Es gibt fünf Möglichkeiten, dataframes in Pandas zusammenzuführen: left, right, outer, inner und cross. In der Dokumentation können Sie nachlesen, wie die anderen vier Methoden funktionieren.  

![Pandas merge](fig/pandas_merge_venn.png)

<div class='alert alert-block alert-success'>

### Aufgabe 2

--> session_04_excercise_2.ipynb

</div>